File: 94-Binary-Tree-Inorder-Traversal.cpp
#include <vector>
#include <set>
#include <stack>

using std::vector;
using std::set;
using std::stack;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        if (root == NULL) {
            return result;
        }
        
        set<TreeNode*> st;
        stack<TreeNode*> stk;
        stk.push(root);
        while (stk.size() > 0) {
            TreeNode* tn = stk.top();
            stk.pop();

            // 第一次visit
            if (st.find(tn) == st.end()) {
                st.insert(tn);
                if (tn->left != NULL) {
                    stk.push(tn);
                    stk.push(tn->left);
                } else {
                    result.push_back(tn->val);
                    if (tn->right != NULL) {
                        stk.push(tn->right);
                    }
                }
            // 第二次visit
            } else {
                result.push_back(tn->val);
                if (tn->right != NULL) {
                    stk.push(tn->right);
                } 
            }
        }

        return result;
    }
};

File: 94-Binary-Tree-Inorder-Traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
	var result []int
	
	if (root == nil) {
		return result
	}

	var stk []*TreeNode
	tn := root;
	for {
		if (tn != nil) {
			stk = append(stk, tn)
			tn = tn.Left
		} else {
			if (len(stk) == 0) {
				return result
			}

			tn = stk[len(stk) - 1]
			stk = stk[:len(stk) - 1]
			result = append(result, tn.Val)
			tn = tn.Right
		}
	}
}